File: backend/app/dm/kpi/schemas.py
"""DM10 閱讀統計 KPI schema（US13 / UCDM13）。"""

from pydantic import BaseModel

from app.core.pagination import PageMetaResponse


class KpiQuery(BaseModel):
    """查詢條件：關鍵字（文件名）、分類。"""

    keyword: str | None = None
    category: str | None = None  # CATEGORY_CODE


class KpiDocItem(BaseModel):
    """逐文件閱讀 KPI 列（FR-002/003）。

    rate 為 None 代表應看=0（無對應閱覽者）：前端顯示「—（無對應閱覽者）」、不計入整體平均。
    """

    doc_id: str
    doc_name: str
    category_code: str
    category_name: str | None
    current_version_no: str | None
    should_see: int
    seen: int
    unseen: int
    rate: float | None  # 0~1；應看=0 → None


class KpiSummary(BaseModel):
    """頂部統計卡（AC3a：整體平均排除應看=0 文件）。"""

    total_docs: int
    overall_rate: float | None  # 全部應看>0 文件之平均閱讀率；無可計文件 → None
    below_50_count: int  # 閱讀率 < 50% 之文件數（僅計應看>0 者）


class KpiListResponse(BaseModel):
    """KPI 儀表板回應：逐文件清單（分頁）+ 統計卡摘要。"""

    data: list[KpiDocItem]
    meta: PageMetaResponse
    summary: KpiSummary

File: frontend/src/dm/kpi/schemas.ts
/** 閱讀統計 KPI（US13 / UCDM13 / DM10）型別（對齊後端 app/dm/kpi/schemas.py）。 */

import type { PagedResult } from "../../hooks/usePagedQuery"

/** 逐文件閱讀 KPI 列。rate 為 null＝應看=0（無對應閱覽者），前端顯示「—」且不計整體平均。 */
export interface KpiDocItem {
  doc_id: string
  doc_name: string
  category_code: string
  category_name: string | null
  current_version_no: string | null
  should_see: number
  seen: number
  unseen: number
  rate: number | null // 0~1
}

/** 頂部統計卡（整體平均排除應看=0 文件）。 */
export interface KpiSummary {
  total_docs: number
  overall_rate: number | null
  below_50_count: number
}

/** KPI 儀表板回應：逐文件清單（分頁）+ 統計卡摘要。 */
export type KpiListResponse = PagedResult<KpiDocItem> & { summary: KpiSummary }

/** 查詢條件（關鍵字＝文件名；分類）。 */
export interface KpiFilters {
  keyword: string
  category: string // '' = 全部
}

export const EMPTY_KPI_FILTERS: KpiFilters = {
  keyword: "",
  category: "",
}
